### PR TITLE
Add support for optional ppx preprocessors

### PR DESCRIPTION
**Pull request by andrewray**
(mainly to support bisect_ppx, but possibly useful elsewhere).

They are specified in a `(preprocess (pps (?ppx)))` stanza.  The
'?' marks them as optional.

A command line argument '--enable-optional-pps=...' takes a comma
separated list of ppx packages to enable.

Note; jbuild.ml; Preprocess_map.filter_optional.  In the
Per_module.Per_module branch, it is possible we could end up with
an empty set of preprocessors (this case is checked in the For_all
branch).  If this is a problem, we probably want a String_map.filter_map
function to be added into import.ml.

### CODE REVIEW

File: src/jbuild.ml [hunk 1]
```diff
@@ -248,6 +252,18 @@ module Preprocess = struct
...
+
+  let filter_optional spec pp =
+    let disable_optional pp =
+      if Pp.is_optional pp && not (String_set.mem (Pp.to_string pp) spec) then None
+      else Some(pp)
+    in
+    match pp with
+    | Pps { pps; flags } ->
+      let pps = List.filter_map pps ~f:disable_optional in
+      if pps = [] then No_preprocessing
```
Comment (diml) [inline src/jbuild.ml]: This should be `pps = [] && flags = []`

File: src/jbuild.mli [hunk 2]
```diff
@@ -50,6 +51,9 @@ module Preprocess_map : sig
   val find : string -> t -> Preprocess.t
 
   val pps : t -> Pp.t list
+
+  (* filter out optional pps which have not been enabled *)
+  val filter_optional : String_set.t -> t -> t
```
Comment (diml) [inline src/jbuild.mli]: Let's use a label: `enabled:String_set.t -> ...`

File: bin/main.ml [hunk 3]
```diff
@@ -166,6 +170,14 @@ let common =
                     build a particular $(b,<package>.install) target.|}
         )
   in
+  let enable_optional_pps =
+    Arg.(value
+         & opt (some (list string)) None
+         & info ["enable-optional-pps"] ~docs ~docv:"PACKAGES"
```
Comment (diml) [inline bin/main.ml]: `PACKAGES` is not very clear here, as we usually don't talk about findlib packages with jbuilder. Let's just use `PPX-REWRITERS`

**Comment by diml** [comment]
Thanks. It seems fine to me to keep a map with entries mapping to `No_preprocessing`

File: src/jbuild.ml (hunk 1 above)
Comment (andrewray) [inline src/jbuild.ml]: Does it really make sense to have flags but no actual preprocessor?  Would this be for the driver?

**Comment by andrewray** [comment]
I've added logic to deal with the issues around flags for optional ppx rewriters, and also command line options to add ppx's globally (except where a preprocess action stanza is used).

**Comment by diml** [comment]
I prefer to have the global pps as a separate PR. Moreover, I don't think it should go in until we can chain pre-processors and so use both an `action` and `pps` preprocessors on the same file. Otherwise it feels a bit like a half-implemented feature.

**Comment by andrewray** [comment]
global ppxs removed.

**Comment by diml** [comment]
Thanks, I'll merge this after the 1.0.0 release

**Comment by aantron** [comment]
@diml, for planning purposes in other projects, could you give a very rough ETA (e.g. 1 month, 3 months, 6 months, 1 year) for 1.1.0 or when this will be released? We won't try to hold Jbuilder to a timeline. It's just so we know if it's worth working around now, or 1.1.0 will be released so soon that it's better to wait for the release.

**Comment by diml** [comment]
So I'm away next week until the 11th. I'm planning to do the 1.0.0 when I get back. After that I think we can do minor releases fairly often. I'd expect the 1.1.0 to come no more than a month later as there are already a lot of new features ready. I just don't want to keep adding new features and postponing the 1.0.0 for now.

**Comment by aantron** [comment]
Thanks!

**Comment by rgrinberg** [comment]
IMO the value of this feature is questionable. It doesn't really make it easier to use bisect because making preprocessors optional is easy enough with the OCaml syntax for jbuild files. And I can't imagine any other use case for this.

**Comment by aantron** [comment]
I don't know if this PR is the right solution, however here is a `jbuild` file in OCaml syntax for making Bisect_ppx optional:

https://github.com/ocsigen/lwt/blob/0e3ade3b740d2b267dc6fe893ce7e656a6d314d6/src/core/jbuild

- I wouldn't say it's that easy – this file is much more difficult to read that a regular `jbuild` file.
- If I want to have this in multiple subdirectories, do I need to convert all the `jbuild` files? Will they have duplicate code? How do I share code between `jbuild` files? Do I need to write a `jbuild`-generating build step?
- The Jbuilder manual [mentions](http://jbuilder.readthedocs.io/en/latest/jbuild.html#ocaml-syntax):

    > It is not clear whether the OCaml syntax will be supported in the long term as it doesn’t work well with incremental builds. It is possible that it will be replaced by just an `include` stanza where one can include a generated file.

**Comment by lpw25** [comment]
I don't know whether this has been suggested before, but couldn't you always run `ppx_bisect` and then pass it an option to tell it whether to do anything. If this option were set through a global variable of some sort, and these variables were settable from the command line, wouldn't that give you the same experience without needing hard-coded support for optional ppxs. It also seems more general, for example allowing you to turn tests on/off: which can't be done with an optional ppx since the ppx still needs to run in both cases (to remove the tests when they are disabled).

**Comment by aantron** [comment]
> ...couldn't you always run ppx_bisect and then pass it an option to tell it whether to do anything...

There is some discussion around here: https://github.com/janestreet/jbuilder/issues/57#issuecomment-308687249.

**Comment by diml** [comment]
Looking at this again, it seems to me that #136 is the right solution to this problem.

`ppx_bisect` could have two implementations:
- a default one that would do nothing and declare no runtime dependencies
- a `coverage` one, that would be the real one

Then, to enable coverage one would do:

```sh
$ jbuilder build -pp-variants coverage
```

where `-pp-variants coverage` would have the effect to enable the `coverage` variant for all ppx rewriters. This would work nicely for things such as inline tests as well, as when dropping tests we wouldn't have to add the runtime dependency.

**Comment by rgrinberg** [comment]
@diml I think bisect already supports pretty much what you're saying through a runtime flag. It uses a shallow identity mapper to implement the "do nothing" variant. @aantron points out that this requires all packages to take on a false dependency and I'm inclined to agree with him that it makes the support partial when compared to other build systems.

**Comment by aantron** [comment]
Furthermore, if one wants to avoid the false dependency, one has to edit `opam` files manually for release. I do this for Lwt. I've heard it mentioned by others, that they don't want to do it for their projects, so they have their `jbuild` file(s) in OCaml syntax.

Before Bisect_ppx 1.3.0, Lwt also had `jbuild` in OCaml syntax, in part for this reason.

**Comment by diml** [comment]
Hmm, I see. I was thinking of dependencies at the library level, but you are thinking of dependencies at the package level. Indeed, with my solution we must always depend on the ppx_bisect package. Although not necessarily the whole ppx_bisect package: the dummy implementation could be distributed as a small standalone package.

I suppose if we do want 0 dependencies when coverage is not enable and we want to keep the syntax light, we don't have much choice. Unless someone has another idea, we should just go ahead with this PR.

**Comment by mseri** [comment]
What is going on with this? It would be great indeed to use something like it to enable/disable ppx_bisect at build time, instead of relying on manual interventions or ugly autogenerations

**Comment by rgrinberg** [comment]
From my perspective this is a non solution. Because I don't want to be editing every single jbuild file I ever wrote just to gain bisect support.

**Comment by edwintorok** [comment]
I think that having optional ppx support in jbuilder is a good starting point, and would remove some of the boilerplate in using `bisect_ppx`.
To avoid changing all the jbuild files could the idea be extended to jbuilder workspaces as well? Then you'd just drop a workspace file wherever you want to enable coverage builds.

**Comment by mseri** [comment]
That sounds better, indeed.

But at the moment bisect support requires changing all the jbuild files anyways, and having something more natural than autogenerating jbuild files or manually editing them each time, would already be a good start.

**Comment by rgrinberg** [comment]
It might be a good start for now, but it would be just cruft down the road. If the optional preprocessing has no use other than enabling/disabling bisect. I also don't see why only preprocessors need to be optional this way. Seems like libraries, module names, flags, etc. should all work with mechanism if we find it desirable. From my experience with oasis, such a "flags" feature did prove to be useful, and specializing to preprocessors does not seem warranted.

**Comment by mseri** [comment]
Fair point, I completely agree with what you say. I think a solution is still needed for the coverage. Would ypu consider the extension of the workspaces suggested by @edwintorok as a better approach?

**Comment by rgrinberg** [comment]
I support the ability of jbuilder automatically configuring coverage for a
build because doing so is easy from an implemention perspective and
requires no boilerplate from users.

I don't see why enabling bisect via workspaces is better than say a command
line flag for example. But I have nothing against it either.

**Comment by aantron** [comment]
If any modifications or restructuring of Bisect are needed, please let me know. We'll be happy to adapt the tool for a well-factored build process.

**Comment by mseri** [comment]
@rgrinberg that is great. Last time there was a discussion on bisect I had understood that making jbuilder deal with it was out of the question

**Comment by rgrinberg** [comment]
Yes, hardcoding it to be only usable for bisect is a no-no. But I'm
thinking perhaps adding a flag (or a workspace config) to allow to add
preprocessors/libraries to libraries/executables. Preferably it would be
something that is useful in other circumstances as well. Though if not, we
could always get rid of it later. At least users won't have to change their
jbuild files.

**Comment by rgrinberg** [comment]
I'm closing this PR as we have a much better story for how to implement bisect support and profile support could subsume this use case as well.

**Pull request closed without merging**